### ISSUE TITLE
Rethink `New` function API

### ISSUE THREAD

**Issue by UlisseMini** [bwmarrin/discordgo#649]
# Problem
The `New` function does not express its intent in its type. it is also very unintuitive to use.
```go
func New(args ...interface{}) (s *Session, err error)
```
The type signature tells the user nothing, i think this API was picked because of different ways to authenticate. I think this is a bad idea and it is more intuitive to use a struct.

# Solution
Add a new more intuitive struct based API. something along the lines of this
```go
dg, err := discordgo.WithAuth {
    Email: "foo@bar.com",
    Pass: "foobar",
    Token: "abc123",
}.Verify()
```
I'm using a method here to prevent me from having to type in`discordgo` again. you could also do it like this
```go
dg, err := discordgo.New(discordgo.WithAuth{
    Email: "foo@bar.com",
    Pass: "foobar",
    Token: "abc123",
})
```
In my opinion both are better then the current API, and could be added while remaining backwards compatible.

If this gets approved and i have time i'll implement this (it should not be that hard).

**Comment by iopred** [comment]
I like where this is heading, an API where you can pass in a WithAuth
option, or WithToken option sounds great to me

**Comment by diamondburned** [comment]
-1, You don't always need an email, a password and a token. I think the current system is fine as it is.

Luckily, since the current `New` function takes in an `interface{}`, you can add this without breaking, and I'm all up for that.

**Comment by UlisseMini** [comment]
@diamondburned If you don't specify a field then it will be set to its zero value, exactly what we want.

for example, not specifying the `Bot` field would leave it to a default of `false`
not specifying an email and password would leave them to `""` and we could check for that in the `New` / `Verify` functions

**Comment by bwmarrin** [comment]
Personally, if we're to break the current API on everyone - I'm more in favor of dropping the Email/Pass entirely and just accepting a single token string that we pass along to Discord.  It's a closer match to the goal of just being a wrapper around Discord's API and it's simple and easy to use.

**Comment by UlisseMini** [comment]
I agree with @bwmarrin, it would also make the API much simpler, something like
```go
func New(token string) (*Session, error)
```
would be desirable

**Comment by pollen5** [comment]
You could also remove the error return value entirely in that case, since currently it's only used for paramater validation and fetching token from email/pass, if it takes only token no error would be returned.

**Comment by diamondburned** [comment]
My PR should add something like this without any breaking changes: https://github.com/bwmarrin/discordgo/pull/725
